src/index.js

```diff
@@ -1 +1,34 @@
-export default () => true
+import defaultsDeep from 'lodash.defaultsdeep';
+import pickBy from 'lodash.pickby';
+import makeOkCheck from './make-ok-check';
+import buildOptions from './build-options';
+import resolveUrl from './resolve-url';
+
+class Stipulate {
+
+  constructor(options) {
+    this.baseOptions = options;
+  }
+
+  beforeRequest(url, options) {
+    return [url, options];
+  }
+
+  afterResponse(response) {
+    return response;
+  }
+
+  send(urlString, options, query) {
+    const config = buildOptions(options, this.baseOptions);
+    const url = resolveUrl(urlString, query);
+
+    const requestArguments = this.beforeRequest(url, config);
+    const checkOk = makeOkCheck(config);
+
+    return fetch(...requestArguments)
+      .then(checkOk)
+      .then(this.afterResponse);
```

Comment:
> it really sounds like you're proposing a middle man between send and the hooks that does the exact same thing that's currently happening.

It's a subtle difference for stronger code.

```js
class ThingWithHooks {
  _beforeHook(value) {
    if (this.before && typeof this.before === 'function') {
      return this.before(value);
    }

    return value;
  }

  _afterHook(value) {
    if (this.after && typeof this.after === 'function') {
      return this.after(value);
    }

    return value;
  }

  doSomething(value) {
    var beforeVal = this._beforeHook(value);
    var fooVal = foo(beforeVal);
    return this._afterHook(fooVal);
  }
}
```

As is, they act as passthroughs with the added benefit that `doSomething` doesn't reference the hooks directly. It's a barrier between the library code and the user's code. The isolation is good because the user's code doesn't interact directly with the internal code and vice versa. Either can change without breaking the other.

But say that we don't want a simple passthrough:

```js
class ThingWithHooks {
  _beforeHook(value) {
    doSomethingElseBeforeDoSomething();
    if (this.before && typeof this.before === 'function') {
      return this.before(value);
    }

    return value;
  }

  _afterHook(value) {
   doSomethingElseAfterDoSomething()
    if (this.after && typeof this.after === 'function') {
      return this.after(value);
    }

    return value;
  }

  doSomething(value) {
    var beforeVal = this._beforeHook(value);
    var fooVal = foo(beforeVal);
    return this._afterHook(fooVal);
  }
}
```

I added my own before and after logic that gets executed without changing the `doSomething` function and it doesn't affect the user's code either.

If you go with this approach, you could move the `checkOk` into the `_afterResponse` method which simplifies the `send` function overall and reduces its responsibility. You could do the same with the config and url resolving if you wanted.

I also think it's a lot cleaner of an implementation.